### PR TITLE
Reporters as Plugins

### LINKED ISSUE #3667: Convert Reporters to a v2-API Plugin Type
### 🎛 Description

Refs #3638

🙋 feature request

Reporters (output formats) are semi-modular (a "v0" plugin).  Most are based off of the RSpec JSON format; but there are kludges involved.  Adding a new reporter is not a clear or easy task, but is often requested.

* [x] Make a Reporter v2-API plugin type
  * [x] Should support one-shot (wait-for-everything) mode
  * ~Should support incremental (progessive) mode~ (deferred to #4144) 
* Clean up existing reporters as needed
* Convert existing reporters into plugins
  * [ ] automate
  * [ ] cli #5069
  * [ ] json
  * [ ] json-automate
  * [ ] json-min #5062
  * [ ] junit #5084
  * [ ] yaml
* [x] Document how to make a new reporter

### 🌍 InSpec and Platform Version

3.0.61+

No work has started as of Dec 11; we are very interested in what people would like in terms of the API.

### PR DESCRIPTION
**Pull request by clintoncwolfe**
## Description

Makes it possible to create a custom reporter and use it with InSpec. Includes documentation and tests.

Notably, this does not introduce a "streaming" API for reporters, per #4144 - that has some fundamental barriers related to how InSpec reconciles RSpecExamples back into Controls, not to mention the sequencing of when reporters are called at all.  That could be introduced later as an optional extension API.

I have not yet converted the  existing reporters into plugins, though that should be straightforward. I was holding off on doing that for at least the Automate ones, as I believe they are being worked on.

## Related Issue

Partially implements #3667 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: lib/inspec/config.rb
Comment (james-stocks) [inline lib/inspec/config.rb]: In the spirit of naming things to reflect current implementation, this could be named `inspec_reporters`.
Naming these `unmigrated`  might suggest that migration is a thing that happens somewhere inside the current InSpec code. 
The comment on the line above leaves a record that these need migrated at some point (we should make sure there's a ticket for doing this as well).

**Comment by james-stocks** [comment]
I'm curious what happens if a user's plugin name clashes with one we ship (after we migrate current reporters to plugins).
Like if a user wants a YAML reporter (lol) named `yaml` and writes one for themself, then the next version of InSpec happens to ship a `yaml` reporter plugin. If this just results in an error, it may be reasonable to ask users to rename their plugins e.g. call theirs `yaml_company`

File: lib/inspec/reporters.rb
Comment (clintoncwolfe) [inline lib/inspec/reporters.rb:34]: A plugin naming collision will throw an error if there is a non-unique name loaded. So if a user installs an `inspec-my-yaml-reporter` that implements a `reporter: yaml` activator (and we've migrated our own YAML reporter), then this line will throw an exception because it expects to find a unique result.

Comment (james-stocks) [reply lib/inspec/reporters.rb:34]: Cool, thanks. It seems fair that we require users to rename their plugin (or move over to using the one we ship, if they do the same thing) :+1:

File: lib/inspec/plugin/v2/plugin_types/reporter.rb
Comment (Schwad) [inline lib/inspec/plugin/v2/plugin_types/reporter.rb:61]: Nice way to enforce render

**Review by Schwad — COMMENTED** [review]
I've had a good read through, found this an informative read. I'm new enough and this PR is deep enough that I can't provide a tremendous amount of architectural feedback (yet). However tomorrow AM I'm planning on having a crack at building my own reporter against this branch to test things out as well. Thanks for this!

**Comment by james-stocks** [comment]
@mjingle - this PR has some docs changes, can you review please?

File: test/helper.rb
Comment (james-stocks) [inline test/helper.rb]: This got merged into master already so can be removed from this PR

File: test/unit/plugin/v2/api_reporter_test.rb
Comment (james-stocks) [inline test/unit/plugin/v2/api_reporter_test.rb]: Small typo `definition`

File: test/unit/plugin/v2/api_reporter_test.rb
Comment (james-stocks) [inline test/unit/plugin/v2/api_reporter_test.rb:13]: I'm maybe not thinking in Ruby here, but isn't `Inspec::Plugin::V2::PluginBase` the superclass and `Inspec::Plugin::V2::PluginType::Reporter` the subclass?

**Comment by james-stocks** [comment]
I wanted to test this manually before approving.
When I try to manually create a plugin in my home folder it doesn't seem to get detected.

The folder I created was:

```
jstocks@phobos inspec % tree ~/.inspec/plugins
/Users/jstocks/.inspec/plugins
└── inspec-jimmy
    └── lib
        └── inspec-jimmy
            ├── plugin.rb
            └── reporter.rb
```

`plugin.rb` contains:

```
module InspecPlugins
  module JimmyReporter
    class Plugin < ::Inspec.plugin(2)
      plugin_name :'inspec-jimmy-reporter'
      reporter :"jimmy" do
        require_relative "reporter.rb"
        InspecPlugins::JimmyReporter::ReporterImplementation
      end
    end
  end
end
```

The exception is like it's not even detected at all:

```
jstocks@phobos inspec % bundle exec inspec exec ../../profiles/inputs --reporter=jimmy
bundler: failed to load command: inspec (/Users/jstocks/code/bundles/inspec/ruby/2.6.0/bin/inspec)
NotImplementedError: 'jimmy' is not a valid reporter type.
  /Users/jstocks/code/inspec/lib/inspec/config.rb:362:in `block in validate_reporters!'
  /Users/jstocks/code/inspec/lib/inspec/config.rb:361:in `each'
  /Users/jstocks/code/inspec/lib/inspec/config.rb:361:in `validate_reporters!'
  /Users/jstocks/code/inspec/lib/inspec/config.rb:474:in `finalize_parse_reporters'
  /Users/jstocks/code/inspec/lib/inspec/config.rb:430:in `finalize_options'
  /Users/jstocks/code/inspec/lib/inspec/config.rb:65:in `initialize'
  /Users/jstocks/code/inspec/lib/inspec/base_cli.rb:249:in `new'
  /Users/jstocks/code/inspec/lib/inspec/base_cli.rb:249:in `config'
  /Users/jstocks/code/inspec/lib/inspec/cli.rb:282:in `exec'

```

File: test/unit/plugin/v2/api_reporter_test.rb
Comment (clintoncwolfe) [reply test/unit/plugin/v2/api_reporter_test.rb:13]: "The superclass" here meaning "the superclass we are creating for the new plugin type" - that is, the class under test at line 6.

File: docs/dev/plugins.md
Comment (mjingle) [inline docs/dev/plugins.md:331]: @cwolfe This new informative documentation is in a dev doc. Does this need to be moved to something more public-facing?

Comment (aaronlippold) [reply docs/dev/plugins.md:331]: Yes, it will land in the Plugins section I would guess

Comment (clintoncwolfe) [reply docs/dev/plugins.md:331]: It is documentation for developers, and is linked from the public-facing Plugins page. I wasn't planning on moving it, this is just adding a section to an existing doc.

Comment (mjingle) [reply docs/dev/plugins.md:331]: @clintoncwolfe Oh! I had forgotten/assumed dev docs were internal only similar to other chef repos. (That link is good to know about as we migrate the docs.) Thank you for clarifying. With that context, this new content should stay where it is. :) Copyedits will appear soon.

**Review by mjingle — APPROVED** [review]
This PR is good from a docs perspective. Thank you for the documentation, Clinton! :D

File: docs/dev/plugins.md
Comment (stevendanna) [inline docs/dev/plugins.md]: If this is intended as a user-facing feature, one worry here is that this effectively makes `run_data` and every object reachable from `run_data` part of the public facing API. Are we OK with that? I took a peek of the surface area there and it doesn't seem _too_ bad on the surface, but my guess is that exposing `Inspec::Profile` probably means most of our major data structures are somehow reachable from run_data.

To completely fix the problem probably involves boiling the ocean, but I'm wondering if we want to do a few things:

1) Wrap the run_data hash in a class with accessors before passing it to reporters. 
2) Extend some of this document with the core bits of this structure that we think we can keep stable over time.
2) Add a warning that accessing data other than those documented might result in their reporter breaking in the future?

(note, this is an actual question, I yield to y'alls judgement)

Comment (Schwad) [reply docs/dev/plugins.md]: Are you concerned that, say for example, someone builds a very involved/intricate reporter plugin that expects `run_data` to be a certain shape and format? And then if we modify it at any point in the future we risk breaking the reporter?

Comment (stevendanna) [reply docs/dev/plugins.md]: @Schwad Yup

**Comment by Schwad** [comment]
This was my first time trying to compose an `inspec-plugin`, but I think this issue was a good opportunity. 

There may be edge cases but @clintoncwolfe I can confirm that I followed the guides and using this branch was able to publish an `inspec-pdf` reporter: https://github.com/Schwad/inspec-pdf . I then installed the plugin.

Against the branch `bundle exec inspec exec my_profile --reporter pdf` worked perfectly and generated this file: 
[inspec-report-152732809.pdf](https://github.com/inspec/inspec/files/4587387/inspec-report-152732809.pdf)

👍

File: docs/dev/plugins.md
Comment (james-stocks) [reply docs/dev/plugins.md]: This is a good point @stevendanna 
I think the more general problem is that there is no API into InSpec for plugins to use.

I believe that currently plugin maintainers are responsible for keeping their plugins working if they want to change InSpec version.
That is a. Unfriendly and b. Possibly very risky when it comes to reporting - their reporter plugin may continue to complete but start to miss bits of result data.

Another option is to maintain a versioned schema for run_data's structure and have reporter plugins need to declare what schema version they are intended for. It will still need plugin developers to keep up to date with InSpec changes, but at least their runs will outright fail instead of resulting in bad report data

Comment (clintoncwolfe) [reply docs/dev/plugins.md]: I'm put some time into making an object model for run_data and documenting as I go. I'm keeping it backwards compatible with Hash access so the existing reports should still work with it.

File: docs/dev/plugins.md
Comment (Schwad) [inline docs/dev/plugins.md]: Thanks for taking the time to break out `run_data` in a way that can be documented and versioned as below, I find this really handy

File: test/unit/plugin/v2/api_reporter_test.rb
Comment (james-stocks) [reply test/unit/plugin/v2/api_reporter_test.rb:13]: This example says that it `returns the superclass` but the assertion is `_(klass).must_equal Inspec::Plugin::V2::PluginType::Reporter`

In inheritance terminology (e.g. http://rubylearning.com/satishtalim/ruby_inheritance.html ) `Inspec::Plugin::V2::PluginType::Reporter` is the subclass not the superclass

File: lib/inspec/plugin/v2/plugin_types/reporter.rb
Comment (james-stocks) [inline lib/inspec/plugin/v2/plugin_types/reporter.rb:4]: Since this PR was raised, the base reporter class had some option handling added to it; it would be great to copy that here so that the options will also work with plugin reporters

https://github.com/inspec/inspec/commit/2103b52dbea8224f0c96d02912ec6cc940d27520#diff-bc3d043c1864e7c4d39d9aec04707b34R13

File: docs/dev/plugins.md
Comment (stevendanna) [reply docs/dev/plugins.md]: 💯 You went far beyond what I expected here. Thanks.

File: lib/inspec/run_data.rb
Comment (james-stocks) [inline lib/inspec/run_data.rb:34]: Awesome 👍

File: lib/inspec/run_data.rb
Comment (james-stocks) [inline lib/inspec/run_data.rb:38]: This says that < 1.0.0 is unstable, but the PR is starting with version 0.1.0. Do you intend to bump this to 1.0.0 before we merge this PR?

Comment (clintoncwolfe) [reply lib/inspec/run_data.rb:38]: @james-stocks and I discussed this a bit, and my thoughts are basically that I feel the API is indeed a bit unstable. This is largely because the source run_data Hash is populated variously depending on the profile being ran, and examined variously depending on the core reporter being used. So, it has been a case of discovery on both ends.

Until the core reporters are adapted to use the RunData object, I'm less confident that there won't be needed changes. That may be non-data changes - like adding more Hash-compatibility features.

That doesn't necessitate blocking this PR on converting the core reporters, though. It does mean that we may want to declare the RunData API as experimental - which it is - and that is a [totally normal thing to do with SemVer](https://semver.org/#spec-item-4).

Alternatively, the RunData API work could be split off, that work incorporated first as an internal refactoring with the core Reporters, then this PR could be revisited with a "mature" API to run against.

Comment (james-stocks) [reply lib/inspec/run_data.rb:38]: I'm thinking we can accept this PR as is, with just a little extra explanation for what 0.1.0 means and why we're not 1.0 (yet)

Comment (clintoncwolfe) [reply lib/inspec/run_data.rb:38]: Added an explanation.

**Comment by james-stocks** [comment]
Let's have expeditor bump to 4.19 when this merges

**Review by james-stocks — APPROVED** [review]
Awesome, thanks @clintoncwolfe !

**Comment by kmf** [comment]
watching

**Pull request merged**
